cpu/cc2538: Add Kconfig support

MrKevinWeiss opened the pull request:
### Contribution description

Add kconfig dependency modeling for cc2538 boards based of @leandrolanzieri work.
This adds the support for the cc2538 remote based boards and radio.
It also adds some ieee802154 based kconfig tests.

### Testing procedure

Check murdock is green and kconfig tests are run for the `remote-revb`

### Issues/PRs references

MrKevinWeiss commented:
I just want to check the CI output first to see what the state is.  It looks like there is some radio work done but maybe that would be better to add in another PR (especially since I think it would also be testing the boards with radios such as the samr)

MrKevinWeiss commented:
Some discussion was had about how `saul_default` should select things and we reached some conclusion that generally the `HAVE_x` in the board definition should be responsible for selecting the `x` part if `saul_default` is enabled but that will be implemented in a different PR.

leandrolanzieri commented on cpu/cc2538/radio/Kconfig:
```diff
@@ -0,0 +1,30 @@
+    depends on TEST_KCONFIG
+    depends on CPU_FAM_CC2538
+    default MODULE_NETDEV_DEFAULT
+    select MODULE_NETDEV_IEEE802154_SUBMAC if MODULE_NETDEV && !MODULE_CC2538_RF_NETDEV_LEGACY
```
I think this was removed not long ago.
@jia200x would you mind taking a look at this and the netdev + submac part?

leandrolanzieri commented on tests/ieee802154_hal/app.config.test:
```diff
@@ -0,0 +1,17 @@
+# this file enables modules defined in Kconfig. Do not use this file for
+# application configuration. This is only needed during migration.
+#CONFIG_MODULE_AUTO_INIT_NRF802154=n
```
Should we remove?

leandrolanzieri commented on tests/ieee802154_submac/app.config.test:
```diff
@@ -0,0 +1,25 @@
+# this file enables modules defined in Kconfig. Do not use this file for
+# application configuration. This is only needed during migration.
+#CONFIG_MODULE_AUTO_INIT_NRF802154=n
```
Here as well

MrKevinWeiss commented:
finally got it working... ignore the bot

MrKevinWeiss commented:
There we go, so I just removed the submac test as we are not adding netdev for this PR.

fjmolinas commented on makefiles/vars.inc.mk:
```diff
@@ -19,6 +19,7 @@ export MCU                   # The MCU, set by the board's Makefile.include, or
 export INCLUDES              # The extra include paths, set by the various Makefile.include files.
 export CXXINCLUDES           # The extra include paths for c++, set by the various Makefile.include files.
 export NATIVEINCLUDES        # The native include paths, set by the various native Makefile.include files.
+export DEVELHELP             # Set to 1 to enable code in RIOT that does safety checking which is not needed in a production environment but helps in the development process.
```
Hmmm...why is this being exported?

MrKevinWeiss replied:
This is part of the first commit.  It is because the `MODULE_MPU_STACK_GUARD` depends on this so kconfing must be able to check it.

fjmolinas replied:
Can you `target-export-variables` only for Kconfig targets?

MrKevinWeiss replied:
Hmmm, I wonder.  It would be needed for any dependency resolution if `TEST_KCONFIG=1`.

fjmolinas replied:
I would prefer not to globally export something that does not need to be globally exported, once exported its really hard to know what are the users of the exported variable, and changes to it become harder.

leandrolanzieri replied:
We should carefully check which of all the Kconfig targets may need it.
As an alternative I'm thinking to have an equivalent Kconfig symbol for develhelp, and set it using https://github.com/RIOT-OS/RIOT/pull/16052

MrKevinWeiss replied:
So would there be a:
```
ifeq (1, $(DEVELHELP))
  RIOT_CONFIG_DEVELHELP=y
endif
```

or something to passthrough the config?

MrKevinWeiss replied:
I don't mind getting in #16052 first as it looks about there.

leandrolanzieri replied:
> So would there be a:

Yes, I imagine something like that, did not test it though...

fjmolinas commented on makefiles/vars.inc.mk:
```diff
@@ -19,6 +19,7 @@ export MCU                   # The MCU, set by the board's Makefile.include, or
 export INCLUDES              # The extra include paths, set by the various Makefile.include files.
 export CXXINCLUDES           # The extra include paths for c++, set by the various Makefile.include files.
 export NATIVEINCLUDES        # The native include paths, set by the various native Makefile.include files.
+export DEVELHELP             # Set to 1 to enable code in RIOT that does safety checking which is not needed in a production environment but helps in the development process.
```
```suggestion
export DEVELHELP             # Set to 1 to enable code in RIOT that does safety checking which might not be needed in a production environment but helps in the development process.
```

The message is wrong though, `enabling` MPI stackguard is something you would like to do in production as well I would think...

MrKevinWeiss replied:
I guess we don't need to worry about this if we get the targets identified...

fjmolinas replied:
Never a bad thing to document a variable though, but out of scope if you target-export in deed

MrKevinWeiss commented:
Rebased and fixed the DEVELHELP export (and added some documentation on the variable).

ping @fjmolinas 

To test changes
```
cd tests/ieee802154_hal
export BOARD=remote-revb
make clean && DEVELHELP=0 TEST_KCONFIG=1 make info-modules | grep mpu
#Should be empty
make clean && TEST_KCONFIG=1 make info-modules | grep mpu
#Should have mpu_stack_guard
```

You can repeat without TEST_KCONFIG to verify it is the same with make.

fjmolinas commented:
> Rebased and fixed the DEVELHELP export (and added some documentation on the variable).
> 
> ping @fjmolinas
> 
> To test changes
> 
> ```
> cd tests/ieee802154_hal
> export BOARD=remote-revb
> make clean && DEVELHELP=0 TEST_KCONFIG=1 make info-modules | grep mpu
> #Should be empty
> make clean && TEST_KCONFIG=1 make info-modules | grep mpu
> #Should have mpu_stack_guard
> ```
> 
> You can repeat without TEST_KCONFIG to verify it is the same with make.

Look good to me, IMO you can squash, any remaining concerns @leandrolanzieri?

leandrolanzieri commented:
> any remaining concerns @leandrolanzieri?

None, it looks good to me!

leandrolanzieri reviewed (approved):
ACK!

leandrolanzieri commented:
There seems to be an unrelated issue on the CI

fjmolinas commented:
> There seems to be an unrelated issue on the CI

Seems like some kind of lack of space issue..  re-triggered

MrKevinWeiss commented:
Thanks for the review and handholding :)
